File: backend-spring-boot/src/test/java/no/vicx/backend/calculator/CalculatorGraphQLControllerSpringBootTest.java
package no.vicx.backend.calculator;

import no.vicx.backend.calculator.vm.CalcVm;
import no.vicx.backend.testconfiguration.TestSecurityConfig;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import java.time.LocalDateTime;
import java.util.List;

import static no.vicx.backend.testconfiguration.TestSecurityConfig.AUTH_HEADER_IN_TEST;
import static no.vicx.database.calculator.CalculatorOperation.PLUS;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@Import(TestSecurityConfig.class)
class CalculatorGraphQLControllerSpringBootTest {

    final static String VALID_DELETE_BODY_IN_TEST = """
            {"query": "mutation { deleteCalculations(ids: [1, 2, 3]) }"}""";

    final static String VALID_CREATE_BODY_IN_TEST = """
            {"query": "mutation { createCalculation(firstValue: 1, secondValue: 2, operation: PLUS) {id firstValue secondValue operation result username createdAt} }"}""";

    @Autowired
    MockMvc mockMvc;

    @MockitoBean
    CalculatorSecurityService calculatorSecurityService;

    @MockitoBean
    CalculatorService calculatorService;

    @Test
    void deleteCalculations_givenNoAuthHeader_expectUnauthorized() throws Exception {
        var responseBody = mockMvc.perform(post("/graphql")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(VALID_DELETE_BODY_IN_TEST))
                .andExpect(status().isOk())
                .andReturn()
                .getResponse()
                .getContentAsString();

        assertEquals(
                """
                        {"errors":[{"message":"Unauthorized","locations":[{"line":1,"column":12}],"path":["deleteCalculations"],"extensions":{"classification":"UNAUTHORIZED"}}],"data":{"deleteCalculations":null}}""",
                responseBody
        );

        verify(calculatorSecurityService, never()).isAllowedToDelete(anyList(), any());
        verify(calculatorService, never()).deleteByIds(anyList());
    }

    @Test
    void deleteCalculations_givenIdsForOtherUser_expectForbidden() throws Exception {
        when(calculatorSecurityService.isAllowedToDelete(anyList(), any()))
                .thenReturn(false);

        var responseBody = mockMvc.perform(post("/graphql")
                        .header(HttpHeaders.AUTHORIZATION, AUTH_HEADER_IN_TEST)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(VALID_DELETE_BODY_IN_TEST))
                .andExpect(status().isOk())
                .andReturn()
                .getResponse()
                .getContentAsString();

        assertEquals(
                """
                        {"errors":[{"message":"Forbidden","locations":[{"line":1,"column":12}],"path":["deleteCalculations"],"extensions":{"classification":"FORBIDDEN"}}],"data":{"deleteCalculations":null}}""",
                responseBody
        );

        verify(calculatorSecurityService).isAllowedToDelete(eq(List.of(1L, 2L, 3L)), any());
        verify(calculatorService, never()).deleteByIds(anyList());
    }

    @Test
    void deleteCalculations_givenIsAllowedToDelete_expectTrue() throws Exception {
        when(calculatorSecurityService.isAllowedToDelete(anyList(), any()))
                .thenReturn(true);

        var responseBody = mockMvc.perform(post("/graphql")
                        .header(HttpHeaders.AUTHORIZATION, AUTH_HEADER_IN_TEST)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(VALID_DELETE_BODY_IN_TEST))
                .andExpect(status().isOk())
                .andReturn()
                .getResponse()
                .getContentAsString();

        assertEquals(
                """
                        {"data":{"deleteCalculations":true}}""",
                responseBody
        );

        verify(calculatorService).deleteByIds(List.of(1L, 2L, 3L));
    }

    @ParameterizedTest
    @ValueSource(booleans = {false, true})
    void createCalculation_expectCreated(boolean isAuthenticated) throws Exception {
        var expectedUsername = isAuthenticated ? "user1" : null;
        var expected = new CalcVm(1L, 1L, 2L, PLUS, 3,
                expectedUsername, LocalDateTime.of(2024, 1, 1, 1, 1, 1));

        when(calculatorService.calculate(any(), any())).thenReturn(expected);

        var requestBuilder = post("/graphql")
                .contentType(MediaType.APPLICATION_JSON)
                .content(VALID_CREATE_BODY_IN_TEST);

        if (isAuthenticated) {
            requestBuilder.header(HttpHeaders.AUTHORIZATION, AUTH_HEADER_IN_TEST);
        }

        var responseBody = mockMvc.perform(requestBuilder)
                .andExpect(status().isOk())
                .andReturn()
                .getResponse()
                .getContentAsString();

        String wrappedUsername = isAuthenticated
                ? "\"user1\""
                : null;

        var expectedResponse = String.format("""
                        {"data":{"createCalculation":{"id":"1","firstValue":1,"secondValue":2,"operation":"PLUS","result":3,"username":%s,"createdAt":"2024-01-01T01:01:01"}}}""",
                wrappedUsername);

        assertEquals(
                expectedResponse,
                responseBody
        );
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/calculator/CalculatorGraphQLControllerSpringBootTest.kt
package no.vicx.backend.calculator

import com.ninjasquad.springmockk.MockkBean
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.data.forAll
import io.kotest.data.row
import io.mockk.called
import io.mockk.every
import io.mockk.verify
import no.vicx.backend.calculator.vm.CalcVm
import no.vicx.backend.SecurityTestUtils.AUTH_HEADER_IN_TEST
import no.vicx.backend.SecurityTestUtils.createPrincipalInTest
import no.vicx.database.calculator.CalculatorOperation
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.security.oauth2.server.resource.introspection.OpaqueTokenIntrospector
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDateTime


@SpringBootTest
@AutoConfigureMockMvc
class CalculatorGraphQLControllerSpringBootTest(
    mockMvc: MockMvc,
    @MockkBean(relaxed = true) private val calculatorSecurityService: CalculatorSecurityService,
    @MockkBean(relaxed = true) private val calculatorService: CalculatorService,
    @MockkBean private val opaqueTokenIntrospector: OpaqueTokenIntrospector
) : BehaviorSpec({
    beforeContainer {
        every { opaqueTokenIntrospector.introspect(any()) } returns createPrincipalInTest(listOf("ROLE_USER"))
    }

    Given("deleteCalculations with mocked environment") {

        When("deleteCalculations without bearer") {
            val resultActions = mockMvc.perform(buildRequest(VALID_DELETE_BODY_IN_TEST,false))

            Then("expect Unauthorized") {
                resultActions
                    .andExpect(status().isOk)
                    .andExpect(jsonPath("$.errors[0].message").value("Unauthorized"))

                verify { calculatorSecurityService wasNot called }
                verify(exactly = 0) { calculatorService.deleteByIds(any()) }
            }
        }

        When("deleteCalculations with ids for other user") {
            every { calculatorSecurityService.isAllowedToDelete(any(), any()) } returns false

            val resultActions = mockMvc.perform(buildRequest(VALID_DELETE_BODY_IN_TEST,true))

            Then("expect Forbidden") {
                resultActions
                    .andExpect(status().isOk)
                    .andExpect(jsonPath("$.errors[0].message").value("Forbidden"))

                verify { calculatorSecurityService.isAllowedToDelete(listOf(1L, 2L, 3L), any()) }
                verify(exactly = 0) { calculatorService.deleteByIds(any()) }
            }
        }

        When("deleteCalculations with ids for current user") {
            every { calculatorSecurityService.isAllowedToDelete(any(), any()) } returns true

            val resultActions = mockMvc.perform(buildRequest(VALID_DELETE_BODY_IN_TEST,true))

            Then("expect OK") {
                resultActions
                    .andExpect(status().isOk)
                    .andExpect(jsonPath("$.data.deleteCalculations").value(true))

                verify { calculatorService.deleteByIds(any()) }
            }
        }
    }

    Given("createCalculation") {
        forAll(
            row("authenticated user", true, "user1"),
            row("anonymous user", false, null)
        ) { description, useAuthenticatedUser, expectedUsername ->

            When("createCalculation, $description") {
                every { calculatorService.calculate(any(), any(), any(), any()) } returns expectedResponse(
                    useAuthenticatedUser
                )

                val resultActions = mockMvc.perform(buildRequest(VALID_CREATE_BODY_IN_TEST,useAuthenticatedUser))

                Then("expect OK") {
                    resultActions
                        .andExpect(status().isOk())
                        .andExpect(jsonPath("$.data.createCalculation.id").value(1))
                        .andExpect(jsonPath("$.data.createCalculation.firstValue").value(1))
                        .andExpect(jsonPath("$.data.createCalculation.secondValue").value(2))
                        .andExpect(jsonPath("$.data.createCalculation.operation").value("PLUS"))
                        .andExpect(jsonPath("$.data.createCalculation.result").value(3))
                        .andExpect(jsonPath("$.data.createCalculation.username").value(expectedUsername))
                }
            }
        }
    }
}) {
    companion object {
        fun buildRequest(
            body: String,
            addBearerToken: Boolean
        ): MockHttpServletRequestBuilder = post("/graphql")
            .contentType(MediaType.APPLICATION_JSON)
            .content(body).apply {
                if (addBearerToken)
                    header(HttpHeaders.AUTHORIZATION, AUTH_HEADER_IN_TEST)
            }

        val VALID_DELETE_BODY_IN_TEST: String = """
            {"query": "mutation { deleteCalculations(ids: [1, 2, 3]) }"}
            """.trimIndent()

        val VALID_CREATE_BODY_IN_TEST: String = """
            {"query": "mutation { createCalculation(firstValue: 1, secondValue: 2, operation: PLUS) {id firstValue secondValue operation result username createdAt} }"}
            """.trimIndent()

        fun expectedResponse(addBearerToken: Boolean) = CalcVm(
            id = 1,
            firstValue = 1,
            secondValue = 2,
            operation = CalculatorOperation.PLUS,
            result = 3,
            username = if (addBearerToken) "user1" else null,
            createdAt = LocalDateTime.now()
        )
    }
}